Modules/_posixsubprocess.c

```diff
@@ -677,6 +712,84 @@ subprocess_fork_exec(PyObject* self, PyObject *args)
         cwd_obj2 = NULL;
     }
 
+    if (call_setgroups) {
+#ifdef HAVE_SETGROUPS
+        if (groups_list != Py_None) {
+            Py_ssize_t i;
+            unsigned long gid;
+
+            if (!PyList_Check(groups_list)) {
+                PyErr_SetString(PyExc_TypeError,
+                        "setgroups argument must be a list");
+                goto cleanup;
+            }
+            num_groups = PySequence_Size(groups_list);
+
+            if (num_groups < 0)
+                goto cleanup;
+
+            if (num_groups > MAX_GROUPS) {
+                PyErr_SetString(PyExc_ValueError, "too many groups");
+                goto cleanup;
+            }
+
+            if ((groups = PyMem_RawMalloc(num_groups * sizeof(gid_t))) == NULL) {
+                PyErr_SetString(PyExc_MemoryError,
+                        "failed to allocate memory for group list");
+                goto cleanup;
+            }
+
+            for (i = 0; i < num_groups; i++) {
+                PyObject *elem;
+                elem = PySequence_GetItem(groups_list, i);
+                if (!elem)
+                    goto cleanup;
+                if (!PyLong_Check(elem)) {
+                    PyErr_SetString(PyExc_TypeError,
+                                    "groups must be integers");
+                    Py_DECREF(elem);
+                    goto cleanup;
+                } else {
+                    /* In posixmodule.c UnsignedLong is used as a fallback value
+                    * if the value provided does not fit in a Long. Since we are
+                    * already doing the bounds checking on the Python side, we
+                    * can go directly to an UnsignedLong here. */
+                    gid = PyLong_AsUnsignedLong(elem);
+                    if ((gid == (unsigned long)-1 && PyErr_Occurred() != NULL)
+                            || (gid > INT_MAX)) {
+
+                        Py_DECREF(elem);
+                        PyErr_SetString(PyExc_ValueError, "invalid group id");
+                        goto cleanup;
+                    }
+                    groups[i] = gid;
+                }
+                Py_DECREF(elem);
+            }
+        }
+#else /* HAVE_SETGROUPS */
+        PyErr_SetString(PyExc_ValueError,
+                "setgroups() is not supported on the current platform");
```

Comment:
It is still useful to raise in this situation as it helps debug internal API use problems.

Reply:
Maybe raise something more serious than a `ValueError` then? It's effectively an assertion failure if we got here.

Reply:
From the description in the docs, it looks like `SystemError` is probably appropriate here, since it's basically a bug in the library if it reaches this (or someone is doing something they shouldn't).

I could also use `RuntimeError` here as well, if `SystemError` is not appropriate.

Reply:
IMHO `SystemError` is appropriate. You may raise it via `PyErr_BadInternalCall`.